File: code/com/leetcode/solutions/NO_0013_罗马数字转整数/Solution.js
const RomanNum = {
    'I': 1,
    'V': 5,
    'X': 10,
    'L': 50,
    'C': 100,
    'D': 500,
    'M': 1000
};

/**
 * 思路：顺序遍历字符串进行转换。并比较上一个罗马数字与当前罗马数字大小，若小则为特殊情况需进行抵消
 * @param {*} s 
 */
var romanToInt = function (s) {
    let n = 0, pre = 0, cur = 0;
    for (let i = 0, len = s.length; i < len; i++) {
        n += cur = RomanNum[s[i]];
        if (pre < cur) n -= pre * 2; //上一个数小于当前数，需要进行抵消并减去
        pre = cur;
    }
    return n;
};

File: code/com/leetcode/solutions/NO_0013_罗马数字转整数/Solution.java
package com.leetcode.solutions.NO_0013_罗马数字转整数;

import java.util.HashMap;
import java.util.LinkedHashMap;

/**
 * Created by gang.xu01@hand-china.com on 2019/1/13
 */
public class Solution {

    private final HashMap<String, Integer> format = new HashMap<String, Integer>(){{
        put("M", 1000);
        put("D", 500);
        put("C", 100);
        put("L", 50);
        put("X", 10);
        put("V", 5);
        put("I", 1);
    }};

    /**
     * 解法：按照顺序排列，左小于右则减，左大于右则加
     *
     * @param s
     * @return
     */
    public int romanToInt(String s) {
        int result = 0, l, r = 0;
        if (s.length() == 1) return format.get(s);
        else for (int i = 0; i < s.length() - 1; i++) {
            l = format.get(String.valueOf(s.charAt(i)));
            r = format.get(String.valueOf(s.charAt(i + 1)));
            if (l < r) result -= l;
            else result += l;
            }
        result += r;
        return result;
    }
}
